src/test/java/spoon/test/lambda/LambdaTest.java

```diff
@@ -450,4 +454,17 @@ private void assertIsWellPrinted(String expected, CtLambda<?> lambda) {
 	private CtLambda<?> getLambdaInFooByNumber(int number) {
 		return foo.getElements(new TypeFilter<CtLambda<?>>(CtLambda.class)).get(number);
 	}
+
+	@Test
+	public void testGetDeclarationOnTypeParameterFromLambda() {
+		List<CtTypeParameterReference> listCtTPR  = launcher.getModel().getElements(new TypeFilter<>(CtTypeParameterReference.class));
+
+		for (CtTypeParameterReference typeParameterReference : listCtTPR) {
+			if (!(typeParameterReference instanceof CtWildcardReference) && typeParameterReference.getDeclaration() == null) {
+				System.err.println(typeParameterReference.getSimpleName()+" from parent "+typeParameterReference.getParent(CtClass.class).getPosition()+"  has null declaration");
+				typeParameterReference.getDeclaration();
+				fail();
+			}
```

Comment:
else { would be nice to check that returned type declaration is the correct one. The last place where it is kniwn is AST of JDT compiler.

But is just an idea. I think you can already merge this PR. The tests are fine, the code is better then before and if there is another problem hidden somewher we can fix it later.